Improve HasFiles performance

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### What would you like to see changed/added?

oh-my-posh startup is extremely slow in mounted directories on a slow NAS with lots of files.

Would it be possible/make sense to cache the directory contents to make those `HasFiles` calls much faster?

```
[DEBUG] 07:59:47.508 shell.go:HasFiles:465 → false
[TRACE] 07:59:47.508 shell.go:HasFiles(pom.xml) - 1.113998292s
[TRACE] 07:59:47.508 shell.go:Pwd() - 459ns
[DEBUG] 07:59:47.509 shell.go:HasFiles:465 → false
[TRACE] 07:59:47.509 shell.go:HasFiles(*.rs) - 1.114309625s
[TRACE] 07:59:47.509 shell.go:Pwd() - 208ns
[DEBUG] 07:59:47.515 shell.go:HasFiles:465 → false
[TRACE] 07:59:47.515 shell.go:HasFiles(workspace.json) - 1.120759333s
[TRACE] 07:59:47.515 shell.go:Pwd() - 625ns
[DEBUG] 07:59:47.516 shell.go:HasFiles:465 → false
[TRACE] 07:59:47.516 shell.go:HasFiles(*.dart) - 1.121688875s
[TRACE] 07:59:47.516 shell.go:Pwd() - 417ns
[ERROR] 07:59:47.588 shell.go:RunCommand:592 → exit status 1
[DEBUG] 07:59:47.588 shell.go:RunCommand:594 ↓
    error: current-context must exist in order to minify
    
[TRACE] 07:59:47.588 shell.go:RunCommand(kubectl config view --output yaml --minify) - 1.192552084s
[DEBUG] 07:59:47.588 properties.go:GetBool:22 → display_error: false
[DEBUG] 07:59:49.105 shell.go:HasFiles:465 → false
[TRACE] 07:59:49.105 shell.go:HasFiles(*.go) - 2.710384s
[TRACE] 07:59:49.105 shell.go:Pwd() - 542ns
[DEBUG] 07:59:49.115 shell.go:HasFiles:465 → false
[TRACE] 07:59:49.115 shell.go:HasFiles(angular.json) - 2.720569875s
[DEBUG] 07:59:49.115 shell.go:HasFiles:465 → false
[TRACE] 07:59:49.115 shell.go:HasFiles(*.rb) - 2.720882292s
[TRACE] 07:59:49.115 shell.go:Pwd() - 625ns
[DEBUG] 07:59:49.121 shell.go:HasFiles:465 → false
[TRACE] 07:59:49.121 shell.go:HasFiles(build.gradle.kts) - 1.612632375s
[TRACE] 07:59:49.121 shell.go:Pwd() - 417ns
[DEBUG] 07:59:49.171 shell.go:HasFiles:465 → false
[TRACE] 07:59:49.171 shell.go:HasFiles(Cargo.toml) - 1.662810958s
[TRACE] 07:59:49.171 shell.go:Pwd() - 625ns
[DEBUG] 07:59:50.675 shell.go:HasFiles:465 → false
[TRACE] 07:59:50.675 shell.go:HasFiles(pubspec.yaml) - 3.158883333s
[TRACE] 07:59:50.675 shell.go:Pwd() - 458ns
[DEBUG] 07:59:50.675 shell.go:HasFiles:465 → false
[TRACE] 07:59:50.675 shell.go:HasFiles(host.json) - 4.280339458s
[TRACE] 07:59:50.675 shell.go:Pwd() - 125ns
[DEBUG] 07:59:50.675 shell.go:HasFiles:465 → false
[TRACE] 07:59:50.675 shell.go:HasFiles(*.cs) - 4.280862292s
[TRACE] 07:59:50.675 shell.go:Pwd() - 250ns
[DEBUG] 07:59:50.688 shell.go:HasFiles:465 → false
[TRACE] 07:59:50.688 shell.go:HasFiles(yarn.lock) - 4.294154s
[TRACE] 07:59:50.689 shell.go:Pwd() - 541ns
[DEBUG] 07:59:50.689 shell.go:HasFiles:465 → false
[TRACE] 07:59:50.689 shell.go:HasFiles(*.jl) - 4.294689709s
[DEBUG] 07:59:51.256 shell.go:HasFiles:465 → false
[TRACE] 07:59:51.256 shell.go:HasFiles(local.settings.json) - 580.690417ms
[TRACE] 07:59:51.256 shell.go:Pwd() - 542ns
[DEBUG] 07:59:51.256 shell.go:HasFiles:465 → false
[TRACE] 07:59:51.256 shell.go:HasFiles(*.csx) - 580.98875ms
[TRACE] 07:59:51.256 shell.go:Pwd() - 250ns
[DEBUG] 07:59:52.938 shell.go:HasFiles:465 → false
[TRACE] 07:59:52.938 shell.go:HasFiles(build.sbt) - 3.817226667s
[DEBUG] 07:59:52.938 shell.go:HasFiles:465 → false
[TRACE] 07:59:52.938 shell.go:Pwd() - 500ns
[TRACE] 07:59:52.938 shell.go:HasFiles(nx.json) - 5.422971125s
[DEBUG] 07:59:52.944 shell.go:HasFiles:465 → false
[TRACE] 07:59:52.944 shell.go:HasFiles(pubspec.yml) - 2.268938083s
[TRACE] 07:59:52.944 shell.go:Pwd() - 541ns
[DEBUG] 07:59:52.944 shell.go:HasFiles:465 → false
[TRACE] 07:59:52.944 shell.go:HasFiles(Cargo.lock) - 3.77240425s
[DEBUG] 07:59:52.944 shell.go:HasFiles:465 → false
[TRACE] 07:59:52.944 shell.go:HasFiles(go.mod) - 3.839026708s
[DEBUG] 07:59:52.959 shell.go:HasFiles:465 → false
[TRACE] 07:59:52.959 shell.go:HasFiles(Rakefile) - 3.84349125s
[TRACE] 07:59:52.959 shell.go:Pwd() - 375ns
[DEBUG] 07:59:53.535 shell.go:HasFiles:465 → false
[TRACE] 07:59:53.535 shell.go:HasFiles(*.vb) - 2.278517375s
[TRACE] 07:59:53.535 shell.go:Pwd() - 459ns
[DEBUG] 07:59:53.547 shell.go:HasFiles:465 → false
[TRACE] 07:59:53.547 shell.go:HasFiles(package-lock.json) - 2.858294875s
[TRACE] 07:59:53.547 shell.go:Pwd() - 541ns
[DEBUG] 07:59:54.104 shell.go:HasFiles:465 → false
[TRACE] 07:59:54.104 shell.go:HasFiles(Gemfile) - 1.145005667s
[DEBUG] 07:59:54.117 shell.go:HasFiles:465 → false
[TRACE] 07:59:54.117 shell.go:HasFiles(function.json) - 2.861780416s
[DEBUG] 07:59:54.679 shell.go:HasFiles:465 → false
[TRACE] 07:59:54.679 shell.go:HasFiles(.java-version) - 1.740961167s
[TRACE] 07:59:54.679 shell.go:Pwd() - 667ns
[DEBUG] 07:59:54.679 shell.go:HasFiles:465 → false
[TRACE] 07:59:54.679 shell.go:HasFiles(pubspec.lock) - 1.735427875s
[DEBUG] 07:59:54.679 shell.go:HasFolder:508 → false
[TRACE] 07:59:54.679 shell.go:HasFolder(.dart_tool) - 29.084µs
[DEBUG] 07:59:55.444 shell.go:HasFiles:465 → false
[TRACE] 07:59:55.444 shell.go:HasFiles(.deps.edn) - 765.368792ms
[TRACE] 07:59:55.444 shell.go:Pwd() - 833ns
[DEBUG] 07:59:55.445 shell.go:HasFiles:465 → false
[TRACE] 07:59:55.445 shell.go:HasFiles(package.json) - 1.897929s
[TRACE] 07:59:55.445 shell.go:Pwd() - 208ns
[DEBUG] 07:59:55.463 shell.go:HasFiles:465 → false
[TRACE] 07:59:55.463 shell.go:HasFiles(*.sln) - 1.928563417s
[TRACE] 07:59:55.463 shell.go:Pwd() - 500ns
[DEBUG] 07:59:56.043 shell.go:HasFiles:465 → false
[TRACE] 07:59:56.043 shell.go:HasFiles(project.clj) - 598.546625ms
[TRACE] 07:59:56.043 shell.go:Pwd() - 917ns
[DEBUG] 07:59:56.043 shell.go:HasFiles:465 → false
[TRACE] 07:59:56.043 shell.go:HasFiles(*.js) - 598.611ms
[TRACE] 07:59:56.043 shell.go:Pwd() - 125ns
[DEBUG] 07:59:56.677 shell.go:HasFiles:465 → false
[TRACE] 07:59:56.677 shell.go:HasFiles(build.boot) - 634.288291ms
[TRACE] 07:59:56.677 shell.go:Pwd() - 500ns
[DEBUG] 07:59:56.678 shell.go:HasFiles:465 → false
[TRACE] 07:59:56.678 shell.go:HasFiles(*.slnf) - 1.2148615s
[TRACE] 07:59:56.678 shell.go:Pwd() - 292ns
[DEBUG] 07:59:57.277 shell.go:HasFiles:465 → false
[TRACE] 07:59:57.277 shell.go:HasFiles(*.ts) - 1.233792542s
[TRACE] 07:59:57.277 shell.go:Pwd() - 500ns
[DEBUG] 07:59:57.277 shell.go:HasFiles:465 → false
[TRACE] 07:59:57.277 shell.go:HasFiles(*.java) - 599.813541ms
[TRACE] 07:59:57.277 shell.go:Pwd() - 208ns
[DEBUG] 07:59:58.183 shell.go:HasFiles:465 → false
[TRACE] 07:59:58.183 shell.go:HasFiles(package.json) - 905.470959ms
[TRACE] 07:59:58.183 shell.go:Pwd() - 458ns
[DEBUG] 07:59:58.183 shell.go:HasFiles:465 → false
[TRACE] 07:59:58.183 shell.go:HasFiles(*.class) - 905.869042ms
[TRACE] 07:59:58.183 shell.go:Pwd() - 250ns
[DEBUG] 07:59:58.183 shell.go:HasFiles:465 → false
[TRACE] 07:59:58.183 shell.go:HasFiles(*.csproj) - 1.504960375s
[TRACE] 07:59:58.183 shell.go:Pwd() - 125ns
[DEBUG] 07:59:58.785 shell.go:HasFiles:465 → false
[TRACE] 07:59:58.785 shell.go:HasFiles(.nvmrc) - 602.735667ms
[TRACE] 07:59:58.785 shell.go:Pwd() - 625ns
[DEBUG] 07:59:58.786 shell.go:HasFiles:465 → false
[TRACE] 07:59:58.786 shell.go:HasFiles(*.gradle) - 602.714208ms
[TRACE] 07:59:58.786 shell.go:Pwd() - 209ns
[DEBUG] 07:59:59.362 shell.go:HasFiles:465 → false
[TRACE] 07:59:59.362 shell.go:HasFiles(pnpm-workspace.yaml) - 576.471292ms
[TRACE] 07:59:59.362 shell.go:Pwd() - 459ns
[DEBUG] 07:59:59.362 shell.go:HasFiles:465 → false
[TRACE] 07:59:59.362 shell.go:HasFiles(*.vbproj) - 1.179282583s
[TRACE] 07:59:59.362 shell.go:Pwd() - 125ns
[DEBUG] 07:59:59.939 shell.go:HasFiles:465 → false
[TRACE] 07:59:59.939 shell.go:HasFiles(.pnpmfile.cjs) - 577.522ms
[TRACE] 07:59:59.939 shell.go:Pwd() - 791ns
[DEBUG] 07:59:59.939 shell.go:HasFiles:465 → false
[TRACE] 07:59:59.940 shell.go:HasFiles(*.jar) - 1.153818584s
[TRACE] 07:59:59.940 shell.go:Pwd() - 167ns
[DEBUG] 08:00:00.522 shell.go:HasFiles:465 → false
[TRACE] 08:00:00.522 shell.go:HasFiles(.npmrc) - 582.883291ms
[TRACE] 08:00:00.522 shell.go:Pwd() - 667ns
[DEBUG] 08:00:00.523 shell.go:HasFiles:465 → false
[TRACE] 08:00:00.523 shell.go:HasFiles(*.fs) - 1.160869292s
[TRACE] 08:00:00.523 shell.go:Pwd() - 500ns
[DEBUG] 08:00:01.123 shell.go:HasFiles:465 → false
[TRACE] 08:00:01.123 shell.go:HasFiles(.vue) - 600.983ms
[DEBUG] 08:00:01.124 shell.go:HasFiles:465 → false
[TRACE] 08:00:01.124 shell.go:HasFiles(*.clj) - 1.184859959s
[TRACE] 08:00:01.124 shell.go:Pwd() - 958ns
[DEBUG] 08:00:01.700 shell.go:HasFiles:465 → false
[TRACE] 08:00:01.700 shell.go:HasFiles(*.fsx) - 1.17724475s
[TRACE] 08:00:01.700 shell.go:Pwd() - 708ns
[DEBUG] 08:00:01.700 shell.go:HasFiles:465 → false
[TRACE] 08:00:01.700 shell.go:HasFiles(*.cljc) - 576.046708ms
[DEBUG] 08:00:01.832 shell.go:HasFiles:465 → false
[TRACE] 08:00:01.832 shell.go:HasFiles(*.fsproj) - 131.163666ms
[TRACE] 08:00:01.832 shell.go:Pwd() - 458ns
[DEBUG] 08:00:01.953 shell.go:HasFiles:465 → false
[TRACE] 08:00:01.953 shell.go:HasFiles(global.json) - 121.375625ms
```

--- a/src/platform/cache.go
+++ b/src/platform/cache.go
@@ -2,13 +2,44 @@ package platform
 
 import (
 	"encoding/json"
+	"fmt"
 	"os"
 	"path/filepath"
+	"strconv"
+	"sync"
 	"time"
 )
 
+type Cache interface {
+	Init(home string)
+	Close()
+	// Gets the value for a given key.
+	// Returns the value and a boolean indicating if the key was found.
+	// In case the ttl expired, the function returns false.
+	Get(key string) (string, bool)
+	// Sets a value for a given key.
+	// The ttl indicates how many minutes to cache the value.
+	Set(key, value string, ttl int)
+	// Deletes a key from the cache.
+	Delete(key string)
+}
+
 const (
 	CacheFile = "/omp.cache"
+)
+
+func pid() string {
+	pid := os.Getenv("POSH_PID")
+	if len(pid) == 0 {
+		pid = strconv.Itoa(os.Getppid())
+	}
+	return pid
+}
+
+var (
+	TEMPLATECACHE    = fmt.Sprintf("template_cache_%s", pid())
+	TOGGLECACHE      = fmt.Sprintf("toggle_cache_%s", pid())
+	PROMPTCOUNTCACHE = fmt.Sprintf("prompt_count_cache_%s", pid())
 )
 
 type cacheObject struct {
@@ -90,4 +121,51 @@ func (fc *fileCache) Set(key, value string, ttl int) {
 func (fc *fileCache) Delete(key string) {
 	fc.cache.Delete(key)
 	fc.dirty = true
+}
+
+type commandCache struct {
+	commands *ConcurrentMap
+}
+
+func (c *commandCache) set(command, path string) {
+	c.commands.Set(command, path)
+}
+
+func (c *commandCache) get(command string) (string, bool) {
+	cacheCommand, found := c.commands.Get(command)
+	if !found {
+		return "", false
+	}
+	command, ok := cacheCommand.(string)
+	return command, ok
+}
+
+type TemplateCache struct {
+	Root          bool
+	PWD           string
+	Folder        string
+	Shell         string
+	ShellVersion  string
+	UserName      string
+	HostName      string
+	Code          int
+	Env           map[string]string
+	Var           SimpleMap
+	OS            string
+	WSL           bool
+	PromptCount   int
+	SHLVL         int
+	Segments      *ConcurrentMap
+	SegmentsCache SimpleMap
+
+	initialized bool
+	sync.RWMutex
+}
+
+func (t *TemplateCache) AddSegmentData(key string, value any) {
+	t.Segments.Set(key, value)
+}
+
+func (t *TemplateCache) RemoveSegmentData(key string) {
+	t.Segments.Delete(key)
 }

--- a/src/platform/shell.go
+++ b/src/platform/shell.go
@@ -37,20 +37,6 @@ const (
 	CMD     = "cmd"
 )
 
-func pid() string {
-	pid := os.Getenv("POSH_PID")
-	if len(pid) == 0 {
-		pid = strconv.Itoa(os.Getppid())
-	}
-	return pid
-}
-
-var (
-	TEMPLATECACHE    = fmt.Sprintf("template_cache_%s", pid())
-	TOGGLECACHE      = fmt.Sprintf("toggle_cache_%s", pid())
-	PROMPTCOUNTCACHE = fmt.Sprintf("prompt_count_cache_%s", pid())
-)
-
 type Flags struct {
 	ErrorCode     int
 	PipeStatus    string
@@ -90,20 +76,6 @@ type FileInfo struct {
 	ParentFolder string
 	Path         string
 	IsDir        bool
-}
-
-type Cache interface {
-	Init(home string)
-	Close()
-	// Gets the value for a given key.
-	// Returns the value and a boolean indicating if the key was found.
-	// In case the ttl expired, the function returns false.
-	Get(key string) (string, bool)
-	// Sets a value for a given key.
-	// The ttl indicates how many minutes to cache the value.
-	Set(key, value string, ttl int)
-	// Deletes a key from the cache.
-	Delete(key string)
 }
 
 type HTTPRequestModifier func(request *http.Request)
@@ -168,36 +140,6 @@ type SystemInfo struct {
 	Disks map[string]disk.IOCountersStat
 }
 
-type TemplateCache struct {
-	Root          bool
-	PWD           string
-	Folder        string
-	Shell         string
-	ShellVersion  string
-	UserName      string
-	HostName      string
-	Code          int
-	Env           map[string]string
-	Var           SimpleMap
-	OS            string
-	WSL           bool
-	PromptCount   int
-	SHLVL         int
-	Segments      *ConcurrentMap
-	SegmentsCache SimpleMap
-
-	initialized bool
-	sync.RWMutex
-}
-
-func (t *TemplateCache) AddSegmentData(key string, value any) {
-	t.Segments.Set(key, value)
-}
-
-func (t *TemplateCache) RemoveSegmentData(key string) {
-	t.Segments.Delete(key)
-}
-
 type Environment interface {
 	Getenv(key string) string
 	Pwd() string
@@ -251,23 +193,6 @@ type Environment interface {
 	DebugF(format string, a ...any)
 	Error(err error)
 	Trace(start time.Time, args ...string)
-}
-
-type commandCache struct {
-	commands *ConcurrentMap
-}
-
-func (c *commandCache) set(command, path string) {
-	c.commands.Set(command, path)
-}
-
-func (c *commandCache) get(command string) (string, bool) {
-	cacheCommand, found := c.commands.Get(command)
-	if !found {
-		return "", false
-	}
-	command, ok := cacheCommand.(string)
-	return command, ok
 }
 
 type Shell struct {

--- a/src/platform/shell.go
+++ b/src/platform/shell.go
@@ -281,6 +281,8 @@ type Shell struct {
 	networks  []*Connection
 
 	sync.RWMutex
+
+	lsDirMap ConcurrentMap
 }
 
 func (env *Shell) Init() {
@@ -430,20 +432,37 @@ func (env *Shell) Pwd() string {
 }
 
 func (env *Shell) HasFiles(pattern string) bool {
+	return env.HasFilesInDir(env.Pwd(), pattern)
+}
+
+func (env *Shell) HasFilesInDir(dir, pattern string) bool {
 	defer env.Trace(time.Now(), pattern)
 
-	cwd := env.Pwd()
-	fileSystem := os.DirFS(cwd)
+	fileSystem := os.DirFS(dir)
+	var dirEntries []fs.DirEntry
 
-	matches, err := fs.ReadDir(fileSystem, ".")
-	if err != nil {
-		env.Error(err)
-		env.Debug("false")
-		return false
+	if files, OK := env.lsDirMap.Get(dir); OK {
+		dirEntries, _ = files.([]fs.DirEntry)
+	}
+
+	if len(dirEntries) == 0 {
+		var err error
+		dirEntries, err = fs.ReadDir(fileSystem, ".")
+		if err != nil {
+			env.Error(err)
+			env.Debug("false")
+			return false
+		}
+
+		env.lsDirMap.Set(dir, dirEntries)
 	}
 
 	pattern = strings.ToLower(pattern)
-	for _, match := range matches {
+
+	env.RWMutex.RLock()
+	defer env.RWMutex.RUnlock()
+
+	for _, match := range dirEntries {
 		if match.IsDir() {
 			continue
 		}
@@ -463,20 +482,6 @@ func (env *Shell) HasFiles(pattern string) bool {
 
 	env.Debug("false")
 	return false
-}
-
-func (env *Shell) HasFilesInDir(dir, pattern string) bool {
-	defer env.Trace(time.Now(), pattern)
-	fileSystem := os.DirFS(dir)
-	matches, err := fs.Glob(fileSystem, pattern)
-	if err != nil {
-		env.Error(err)
-		env.Debug("false")
-		return false
-	}
-	hasFilesInDir := len(matches) > 0
-	env.DebugF("%t", hasFilesInDir)
-	return hasFilesInDir
 }
 
 func (env *Shell) HasFileInParentDirs(pattern string, depth uint) bool {

